OTAnalytics/domain/track.py

```diff
@@ -1,33 +1,100 @@
-from dataclasses import dataclass
 from datetime import datetime
 from pathlib import Path
 from typing import Iterable, Optional
 
+from pydantic import BaseModel, Field, validator
 
-@dataclass(frozen=True)
-class Detection:
-    classification: str
-    confidence: float
-    x: float
-    y: float
-    w: float
-    h: float
-    frame: int
+import OTAnalytics.plugin_parser.ottrk_dataformat as ottrk_format
+
+
+class Detection(BaseModel, frozen=True, allow_population_by_field_name=True):
+    """Represents a detection belonging to a `Track`.
+
+    The detection uses the xywh bounding box format.
+
+    Extends from pydantic.BaseModel.
+    IMPORTANT: Instantiation of this class is only possible by using named parameters.
+
+    Raises:
+        ValueError: confidence not in [0,1]
+        ValueError: x < 0
+        ValueError: y < 0
+        ValueError: w < 0
+        ValueError: h < 0
+        ValueError: frame < 0
+        ValueError: track_id < 0
+
+
+    Args:
+        classification (str): class of detection.
+        confidence (float): the confidence.
+        x (float): the x coordinate component of the bounding box.
+        y (float): the y coordinate component of the bounding box.
+        w (float): the width component of the bounding box.
+        h (float): the height component of the bounding box.
+        frame (int): the frame that the detection belongs to.
+        occurrence (datetime): the time of the detection's occurence.
+        input_file_path (Path): absolute path to otdet that the detection belongs to
+        at the time of its creation.
+        interpolated_detection (bool): whether this detection is interpolated.
+        track_id (int): the track id this detection belongs to.
+    """
+
+    classification: str = Field(alias=ottrk_format.CLASS)
+    confidence: float = Field(gt=0, le=1, description="must be between range [0,1]")
+    x: float = Field(ge=0, description="must be greater equal than 0")
+    y: float = Field(ge=0, description="must be greater equal than 0")
+    w: float = Field(ge=0, description="must be greater equal than 0")
+    h: float = Field(ge=0, description="must be greater equal than 0")
+    frame: int = Field(gt=0, description="must be greater than 0")
     occurrence: datetime
     input_file_path: Path
-    interpolated_detection: bool
-    track_id: int
+    interpolated_detection: bool = Field(alias=ottrk_format.INTERPOLATED_DETECTION)
```

Comment:
Why do some fields use aliases and some not?

Reply:
That's a good question! 

First up why I decided to use aliases at all. `INTERPOLATED_DETECTION` in the ottrk file is `interpolated-detection`. In Python, hyphens are not allowed to be used in variable names. Thus, it is spelled using snake_case. I thought it would be neat to use pydantic's serialisation features at some point in the future. At this moment I only use it in our tests. Thus, defining aliases allows us to easily de- or serialise a detection without the need to care about making the decision to use snake_case or hyphen variant of as a key in this use case. 

Now to answer your question, I wasn't thinking ahead. A change in the data format of ottrk might require changes here as well. So, defining aliases for the other attributes might be an appropriate thing to do.

Reply:
We should discuss if we want third party library dependencies in domain and application code. If not, we can move the hyphen vs. underscore in names into a plugin. Our domain code should not know anything about serialisation because it is not part of our domain.